src/main/java/uk/ac/cam/cl/dtg/segue/api/SessionValidator.java

```diff
@@ -0,0 +1,107 @@
+import jakarta.ws.rs.container.ContainerResponseContext;
+import jakarta.ws.rs.container.ContainerResponseFilter;
+import jakarta.ws.rs.container.PreMatching;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.Cookie;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.ext.Provider;
+import org.jboss.resteasy.core.interception.jaxrs.ContainerResponseContextImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.ac.cam.cl.dtg.segue.api.managers.UserAuthenticationManager;
+import uk.ac.cam.cl.dtg.util.PropertiesLoader;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Map;
+
+import static uk.ac.cam.cl.dtg.segue.api.Constants.*;
+
+@Provider
+@PreMatching
+public class SessionValidator implements ContainerRequestFilter, ContainerResponseFilter {
+
+    private static final Logger log = LoggerFactory.getLogger(SessionValidator.class);
+
+    private final UserAuthenticationManager userAuthenticationManager;
+    private final PropertiesLoader properties;
+    private final Integer sessionExpirySeconds;
+    @Context
+    private HttpServletRequest httpServletRequest;
+    @Context
+    private HttpServletResponse httpServletResponse;
+
+    @Inject
+    public SessionValidator(final UserAuthenticationManager userAuthenticationManager, PropertiesLoader properties) {
+        this.userAuthenticationManager = userAuthenticationManager;
+        this.properties = properties;
+        this.sessionExpirySeconds = Integer.parseInt(this.properties.getProperty(SESSION_EXPIRY_SECONDS_DEFAULT));
+    }
+
+    @Override
+    public void filter(ContainerRequestContext containerRequestContext) throws IOException {
+        Cookie authCookie = containerRequestContext.getCookies().get(SEGUE_AUTH_COOKIE);
+        if (authCookie != null) {
+            if (!userAuthenticationManager.isSessionValid(httpServletRequest)) {
+                log.warn("Request made with invalid segue auth cookie - closing session");
+                httpServletRequest.getSession().invalidate();
+                containerRequestContext.abortWith(Response
+                        .status(Response.Status.BAD_REQUEST)
+                        .entity("Authentication cookie is invalid")
+                        .cookie(userAuthenticationManager.createAuthLogoutNewCookie())
+                        .build()
+                );
+            }
+        }
+    }
+
+    @Override
+    public void filter(ContainerRequestContext containerRequestContext, ContainerResponseContext containerResponseContext) throws IOException {
+        Cookie authCookie = containerRequestContext.getCookies().get(SEGUE_AUTH_COOKIE);
+        if (authCookie != null && !isLogoutCookiePresent(httpServletResponse) && wasRequestValid(containerResponseContext)) {
+            try {
+                jakarta.servlet.http.Cookie newAuthCookie = generateRefreshedSegueAuthCookie(authCookie);
```

Comment:
Is there a way that a user could trick the function into refreshing the auth cookie when it shouldn't? I think this is fine otherwise, so if you could double-check my logic here, that would be appreciated...

Reply:
A "logout cookie" is currently defined as a cookie with the SEGUE_AUTH_COOKIE name and a max age of zero. That max age means the browser should immediately remove it; as I understand it, that's the standard method for removing cookies.
If you were to send a request with that cookie attached using something like postman without changing the age, it should be recognised as a logout cookie and not trigger a refresh, assuming it wasn't rejected as invalid.
If you edited the max-age and attached that to a request, that _should_ fail validation.
Are there any scenarios I've not covered there?